transport/transport.go

```diff
@@ -351,9 +353,23 @@ type ConnectOptions struct {
 	Timeout time.Duration
 }
 
+// WithDialer returns a ConnectOptions with the specified dialer function.
+func (copts ConnectOptions) WithDialer(f func(string, time.Duration, <-chan struct{}) (net.Conn, error)) ConnectOptions {
+	copts.dialer = f
+	return copts
+}
+
+// Dial the address using the previously specified dialer or a default dialer.
+func (copts ConnectOptions) Dial(addr string, timeout time.Duration, cancel <-chan struct{}) (net.Conn, error) {
```

Comment:
It sounds strange to have a Dial function for an options object. Probably make it something like:

```
func Dial(copts ConnetionOptions, addr, string, timeout ....) (net.Conn, error) {
   ...
}
```